### PR TITLE
Wrapper script to run healthchecks

### PR DESCRIPTION
**Pull request by muffato**
## Description

The command-lines to run the HCs require some customisation for every division (host, port, database name, etc) but all of those are either already available in the production_reg_conf file, or could be added to another file. My goal is to have a "universal" command-line that works for all divisions.

## Overview of changes

Here I introduce a wrapper script that allows running the HCs with simple commands such as:
```
perl run_healthchecks.pl $COMPARA_REG compara_curr -g ComparaAll
perl run_healthchecks.pl $COMPARA_REG compara_curr -t org.ensembl.healthcheck.testcase.compara.MLSSTagGERPMSA --repair
```

The wrapper will automatically select the correct database locations to run the healthchecks. This means we don't have to change `database.default.properties` any more and can leave the default options there (command-line arguments override the config file).

#### Change #1

Importantly, I had to introduce a new per-division configuration file, which I have named `ensj-healthcheck.json`. It is expected to contain a subset of those keys: `host`, `host2`, `host3` and `secondary.host`. The script will automatically extrapolate the other parameters (port, etc) and unset the missing parameters

## Testing

I have used to run HCs for Pan in e100. I have also tested `org.ensembl.healthcheck.testcase.compara.CheckGenomeDB` for every division since this checks critically needs all the Core databases to be set up (in the right order).

## Notes

I haven't yet changed the PrepareMasterDatabase pipeline, but I can update `RunnableDB/RunJavaHealthCheck.pm` if you like this script.

I don't 100% remember how the pipelineto build CITest works, especially what it does with the `database.default.properties` file (cf ReconfigPipeline.pm and the pipeline_create_commands in the PipeConfig), so would need a refresher from @JAlvarezJarreta on this one.

### CODE REVIEW

**Comment by muffato** [comment]
I can write a similar script for the datachecks as well that would 1) automatically add the production_reg_conf and 2) allow using registry aliases like `compara_curr`

File: scripts/production/run_healthchecks.pl [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    );
+}
+
+# Add the master database if one is available (otherwise defaults to ensj-healthcheck name matching)
+if ($master_db) {
+    my $master_dba = Bio::EnsEMBL::Compara::DBSQL::DBAdaptor->go_figure_compara_dba($master_db);
+    die "Cannot find a database with the alias '$master_db' in the Registry" unless $master_dba;
+    push @params, (
+        '--compara_master.database', $master_dba->dbc->dbname,
+    );
```
Comment (JAlvarezJarreta) [inline scripts/production/run_healthchecks.pl]: Just a picky comment: naming variables `master_db` and `master_dba` might lead to hard-to-find bugs if you misspell the latter, i.e. by mistake you don't add the last `a`.

File: scripts/production/run_healthchecks.pl [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+            "--${key_prefix}host${key_suffix}"    => '""',
+        );
+    }
+}
+
+print "Executing: ", join(" ", $ensj_testrunner, @params, @ARGV), "\n\n";
+
+# Need to change directory because database.default.properties is read from the current directory
+chdir dirname($ensj_testrunner);
+exec($ensj_testrunner, @params, @ARGV);
```
Comment (JAlvarezJarreta) [inline scripts/production/run_healthchecks.pl]: Does this leave you in the same path where you called the script after if finishes?

**Comment by JAlvarezJarreta** [comment]
If you update the script for `RunnableDB/RunJavaHealthCheck.pm`, then there is no need to pass and reconfigure `database.default.properties` in `RunnableDB/BuildMaster/ReconfigPipeline.pm`.

File: scripts/production/run_healthchecks.pl (hunk 1 above)
Comment (carlacummins) [reply scripts/production/run_healthchecks.pl]: I don't agree. This is how I always name my variables and have never struggled to find a bug caused by it.

Comment (muffato) [reply scripts/production/run_healthchecks.pl]: Yes, it's quite standard to name the variables _db, _dba and _dbc

File: scripts/production/run_healthchecks.pl (hunk 2 above)
Comment (muffato) [reply scripts/production/run_healthchecks.pl]: Yes, the `chdir` only applies to the script's process, and does no bubble up to your shell

File: modules/Bio/EnsEMBL/Compara/PipeConfig/Parts/PrepareMasterDatabaseForRelease.pm [hunk 3]
```diff
@@ -251,7 +251,9 @@ sub pipeline_analyses_prep_master_db_for_release {
                 'work_dir'    => $self->o('work_dir'),
                 'testgroup'   => 'ComparaMaster',
                 'output_file' => '#work_dir#/healthcheck.#testgroup#.out',
-                'java_hc_dir' => $self->o('java_hc_dir'),
+                'ensj_conf'   => $self->o('ensj_conf'),
+                'run_healthchecks_exe'  => $self->o('run_healthchecks_exe'),
+                'ensj_testrunner_exe'   => $self->o('ensj_testrunner_exe'),
```
Comment (JAlvarezJarreta) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/Parts/PrepareMasterDatabaseForRelease.pm]:
```
                'run_healthchecks_exe' => $self->o('run_healthchecks_exe'),
                'ensj_testrunner_exe'  => $self->o('ensj_testrunner_exe'),
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/BuildMaster/ReconfigPipeline.pm [hunk 4]
```diff
@@ -92,13 +87,12 @@ sub run {
...
     # Modify the registry configuration file
     $self->_spurt($reg_conf, $content);
-    # All cloned core databases are in the same host, so replace that
-    # information in the Java healthchecks database properties file ('host',
-    # 'host1' and 'host2', and 'port', 'port1' and 'port2')
-    $content = $self->_slurp($java_hc_db_prop);
-    $content =~ s/(^)(host[12]?[ ]*=)[ ]*[\w\.-]+/$1$2 $dst_host/gm;
-    $content =~ s/(^)(port[12]?[ ]*=)[ ]*\d+/$1$2 $dst_port/gm;
-    $self->_spurt($java_hc_db_prop, $content);
+    # And in the compara ensj-healthcheck configuration file
```
Comment (JAlvarezJarreta) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/BuildMaster/ReconfigPipeline.pm]:
```suggestion
    # All cloned core databases are in the same server, so update that information
    # in the compara ensj-healthcheck configuration file
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/BuildMaster/ReconfigPipeline.pm [hunk 5]
```diff
@@ -35,8 +31,6 @@ Requires several inputs:
     'reg_conf'        : full path to the registry configuration file
     'reg_conf_tmpl'   : full path to the registry configuration template file
     'master_db'       : new master database
-    'java_hc_db_prop' : full path to 'database.default.properties' file (used by
-                        Java healthchecks)
```
Comment (JAlvarezJarreta) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/BuildMaster/ReconfigPipeline.pm]:
```
    'ensj_conf'       : full path to 'ensj-healthcheck.json' file (used to run
                        Java healthchecks)
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/RunJavaHealthCheck.pm [hunk 6]
```diff
@@ -27,9 +27,7 @@ Bio::EnsEMBL::Compara::RunnableDB::RunJavaHealthCheck
 
 Runs an EnsEMBL Java Healthcheck (see https://github.com/Ensembl/ensj-healthcheck)
 Requires several inputs:
-    'java_hc_dir' : path to ensj-healthcheck root dir
```
Comment (JAlvarezJarreta) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/RunJavaHealthCheck.pm]:
```
    'run_healthchecks_exe' : path to compara run_healthchecks script
    'ensj_testrunner_exe'  : path to ensj-healthcheck testrunner script
    'ensj_conf'            : path to compara ensj-healthcheck JSON configuration file
```

File: scripts/production/run_healthchecks.pl [hunk 7]
```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env perl
+# Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
```
Comment (JAlvarezJarreta) [inline scripts/production/run_healthchecks.pl]: Not sure if in scripts `=head1 LICENSE` should be added between the shebang line and the copyright.

File: scripts/production/run_healthchecks.pl [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+The script reads these arguments and passes the other ones straight to run-configurable-testrunner.sh
+
+=head2 DATABASE SETUP
+
+=over
+
+=item B<[--url mysql://user[:passwd]@host[:port]/dbname]>
+
+URL of the database to test
```
Comment (JAlvarezJarreta) [inline scripts/production/run_healthchecks.pl]:
```suggestion
URL of the database to test.
```
(just to follow your own style below)

File: scripts/production/run_healthchecks.pl [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    'reg_type=s'                    => \$reg_type,
+    'reg_alias|regname|reg_name=s'  => \$reg_alias,
+    'master_db|compara_master=s'    => \$master_db,
+    'ensj-testrunner=s'             => \$ensj_testrunner,
+    'ensj-json-config=s'            => \$ensj_json_config,
+    'repair'                        => \$repair,
+);
+
+unless ($url or ($reg_conf and $reg_alias)) {
+    print "\nNeither --url nor --reg_conf and --reg_alias are defined. The URL should be something like mysql://ensro\@compara1:3306/kb3_ensembl_compara_59\nEXIT 2\n\n";
```
Comment (JAlvarezJarreta) [inline scripts/production/run_healthchecks.pl]: Out of curiosity, why the `EXIT 2` bit?

File: scripts/production/run_healthchecks.pl (hunk 7 above)
Comment (muffato) [reply scripts/production/run_healthchecks.pl]: The convention for scripts is that the license should not be in a `-head` section but in `#` comments. See https://www.ebi.ac.uk/seqdb/confluence/display/ENS/Ensembl+Software+License

File: scripts/production/run_healthchecks.pl (hunk 9 above)
Comment (muffato) [reply scripts/production/run_healthchecks.pl]: Unnecessary copy-paste. I have restored this to `exit 1`

Comment (JAlvarezJarreta) [reply scripts/production/run_healthchecks.pl]: Sorry, I meant in the `print` command. Although I assume if you have changed below, in the printed string it should say `EXIT 1` too?
```suggestion
    print "\nNeither --url nor --reg_conf and --reg_alias are defined. The URL should be something like mysql://ensro\@compara1:3306/kb3_ensembl_compara_59\nEXIT 1\n\n";
```

Comment (muffato) [reply scripts/production/run_healthchecks.pl]: Sorry I didn't see it, the line was too long :D !
I'll remove the `EXIT`, we usually don't print it

```suggestion
   print "\nNeither --url nor --reg_conf and --reg_alias are defined. The URL should be something like mysql://ensro\@compara1:3306/kb3_ensembl_compara_59\n\n";
```

**Comment by muffato** [comment]
I have used the new script `run_healthchecks.pl` in e100 production (for GRCh37 and Pan) but I have not used the updated `PrepareMasterDatabaseForRelease` pipeline. I feel like I should merge this into the master branch, but release/100. What do you think ?

**Comment by JAlvarezJarreta** [comment]
I know you are much more proficient than I am with `git`, but rebasing from e100 to master can be quite tedious. I would leave it as it is.

**Comment by muffato** [comment]
No need to rebase it, just change the target branch at the top of this page

**Pull request merged**